Src/PerformanceCollector/Shared/Implementation/QuickPulse/QuickPulseServiceClient.cs

```diff
@@ -0,0 +1,228 @@
+    using System.IO;
+    using System.Linq;
+    using System.Net;
+    using System.Runtime.Serialization.Json;
+    using System.Web;
+
+    using Microsoft.ManagementServices.RealTimeDataProcessing.QuickPulseService;
+
+    /// <summary>
+    /// Service client for QPS service.
+    /// </summary>
+    internal sealed class QuickPulseServiceClient : IQuickPulseServiceClient
+    {
+        private const string XMsQpsSubscribedHeaderName = "x-ms-qps-subscribed";
+
+        private const int RetryCount = 3;
+        
+        private readonly string instanceName;
+
+        private readonly string version;
+
+        private readonly TimeSpan timeout = TimeSpan.FromSeconds(3);
+        
+        public QuickPulseServiceClient(Uri serviceUri, string instanceName, string version, TimeSpan? timeout = null)
+        {
+            this.ServiceUri = serviceUri;
+            this.instanceName = instanceName;
+            this.version = version;
+            this.timeout = timeout ?? this.timeout;
+        }
+
+        public Uri ServiceUri { get; }
+
+        public bool? Ping(string instrumentationKey, DateTime timestamp)
+        {
+            MemoryStream bodyStream = this.WritePingData(timestamp);
+
+            var path = string.Format(CultureInfo.InvariantCulture, "ping?ikey={0}", instrumentationKey);
+            HttpWebResponse response = this.SendRequest(WebRequestMethods.Http.Post, path, bodyStream, true);
+            if (response == null)
+            {
+                return null;
+            }
+            
+            return ProcessResponse(response);
+        }
+        
+        public bool? SubmitSamples(IEnumerable<QuickPulseDataSample> samples, string instrumentationKey)
+        {
+            MemoryStream bodyStream = WriteSamples(samples, instrumentationKey);
+            
+            var path = string.Format(CultureInfo.InvariantCulture, "post?ikey={0}", HttpUtility.UrlEncode(instrumentationKey));
+            HttpWebResponse response = this.SendRequest(WebRequestMethods.Http.Post, path, bodyStream, false);
+            if (response == null)
+            {
+                return null;
+            }
+
+            return ProcessResponse(response);
+        }
+
+        private static bool? ProcessResponse(HttpWebResponse response)
+        {
+            bool isSubscribed;
+            if (!bool.TryParse(response.GetResponseHeader(XMsQpsSubscribedHeaderName), out isSubscribed))
+            {
+                return null;
+            }
+
+            return isSubscribed;
+        }
+
+        private static double Round(double value)
+        {
+            return Math.Round(value, 4, MidpointRounding.AwayFromZero);
+        }
+
+        private MemoryStream WritePingData(DateTime timestamp)
+        {
+            var bodyStream = new MemoryStream();
+
+            var dataPoint = new MonitoringDataPoint
+            {
+                Version = this.version,
+                //InstrumentationKey = instrumentationKey,
```

Comment:
Right now the service doesn't expect the instrumentation key to be set in the body of Ping request, it takes it from the query parameter. This is likely to be adjusted in future iterations, but as long as this weirdness exists, I'd rather keep it here as a reminder of what's going on.